src/core/service.c

```diff
@@ -2695,6 +2698,13 @@ static void service_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                                 /* Fall through */
 
                         case SERVICE_RUNNING:
+                                if (s->type == SERVICE_NOTIFY) {
+                                        /* Only enter running through a notification */
+                                        if (f != SERVICE_SUCCESS)
+                                                service_enter_signal(s, SERVICE_FINAL_SIGTERM, f);
+                                        break;
+                                }
+
                                 service_enter_running(s, f);
```

Comment:
Note that service_enter_running() will actually not always place the unit in the SERVICE_RUNNING state, but do this only after checking that everything is OK. My guess is that the block in that function, that looks for notifications messages received so far, should also check for NOTIFY_READY and if we didn't get it yet, should not activate SERVICE_RUNNING yet, but exit...